design/ceph/osd-migration.md

```diff
@@ -0,0 +1,234 @@
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the OSD should be migrated in case the user changes the backend store type in the spec (`spec.Storage.osdStore.type`).
+
+### New OSDs
+
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the OSDs.
+
+#### OSD Prepare
+
+- The `ceph-volume prepare` command is updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+    - RAW MODE
+        ```
+        ceph-volume raw prepare <OSD_STORE_FLAG> --data /dev/vda
+        ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm prepare <OSD_STORE_FLAG> --data /dev/vda
+      ```
+
+#### OSD Create
+
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+
+    - RAW MODE
+      ```
+      ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs
+      ```
+
+    - LVM MODE
+      ```
+      ceph-volume lvm activate --no-systemd "$OSD_STORE_FLAG" "$OSD_ID" "$OSD_UUID"
+      ```
+
+**NOTE**: ceph-volume `activate` command doesn't require `OSD_STORE_FLAG` to be passed as an argument. It auto detects the backend store that was used during the OSD `prepare`
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  osdStore:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### API
+
+```go
+// OSDReplaceInfo represents an OSD that needs to replaced
+type OSDReplaceInfo struct {
+	ID   int    `json:"id"`
+	Path string `json:"path"`
+	Node string `json:"node"`
+}
+
+// ReplaceQueue contains the list of OSDs to be replaced one at a time
+type ReplaceQueue struct {
+	OSDs []OSDReplaceInfo
+}
+
+// OSDReplaceConfig contains the status of the OSD replacement
+type OSDReplaceConfig struct {
+	OSDList              *ReplaceQueue   `json:"osdList,omitempty"`
+	IsPgHealthy          bool            `json:"isPgHealthy,omitempty"`
+	PgHealthMessage      string          `json:"pgHealthMessage,omitempty"`
+	LastReplacedOSD      *OSDReplaceInfo `json:"lastReplacedOSD,omitempty"`
```

Comment:
Instead the last replaced OSD, I'm thinking we need to store the OSDInProgress. The sequence would be:
- Remove an OSD from the list and set it to OSDInProgress
- The operator starts the replacement progress
- After the OSD is replaced and its OSD deployment is running again, reset the OSDInProgress to blank. This can be reset as soon as the deployment is running and before the PGs are active+clean.

Reply:
This way if the operator is restarted in the middle of a replace, it can attempt to restart the replacement, for example if the osd prepare job was never created or failed.

Reply:
@BlaineEXE also suggested something similar. But he suggested it as a new option along with the `LastReplaceOSD`.  The LastReplaceOSD and LastReplaceOSDTimeStamp might be useful when an pgs are taking a lot of time to be `active+clean`.  Users can know which OSD was last replaced and the time elapsed since then.

Reply:
So the update flow can be like: 
- Remove an OSD from the list and set it to OSDInProgress
- The operator starts the replacement progress
- After the OSD is replaced and its OSD deployment is running again, reset the OSDInProgress to blank. (This can be reset as soon as the deployment is running and before the PGs are active+clean)
- Update the `LastReplaceOSD` to `OSDInProgress` and `LastReplacedOSDTimestamp` to current time.


Suggestions?

Reply:
marking this as resolved since the design doc is updated now.